lib/collections.c

```diff
@@ -129,3 +131,103 @@ int64_t conch_checkout(Conch* conch)
 bool conch_is_available(Conch* conch) { return conch->conch_available; }
 
 void conch_free(Conch* conch) { free(conch); }
+
+// Dynamic Array ADT
+typedef struct List
+{
+    size_t length;
+    size_t max_length;
+    void** data;
+} List;
+
+List* list_init(void)
+{
+    List* list = malloc(sizeof(List));
+    if (list == NULL)
+    {
+        fprintf(stderr, "Error: list_init: malloc failed\n");
+        exit(EXIT_FAILURE);
+    }
+    list->length = 0;
+    list->max_length = INITIAL_LIST_MAX_LENGTH;
+    list->data = malloc(INITIAL_LIST_MAX_LENGTH * sizeof(void*));
+    if (list->data == NULL)
+    {
+        fprintf(stderr, "Error: list_init: malloc failed\n");
+        exit(EXIT_FAILURE);
+    }
+    return list;
+}
+
+void* list_get_at(List* list, size_t index)
+{
+    if (index >= list->length)
+    {
+        fprintf(stderr, "Error: list_get_at: index out of bounds\n");
+        exit(EXIT_FAILURE);
+    }
+    return list->data[index];
+}
+
+void list_set_at(List* list, size_t index, void* item)
+{
+    if (index >= list->length)
+    {
+        fprintf(stderr, "Error: list_set_at: index out of bounds\n");
+        exit(EXIT_FAILURE);
+    }
+    list->data[index] = item;
+}
+
+static void list_resize(List* list)
+{
+    if (list->length == list->max_length)
+    {
+        list->max_length *= 2;
+        list->data = realloc(list->data, list->max_length * sizeof(void*));
+        if (list->data == NULL)
+        {
+            fprintf(stderr, "Error: list_resize: realloc failed\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+    else if (list->length * 4 <= list->max_length)
+    {
+        list->max_length /= 2;
+        list->data = realloc(list->data, list->max_length * sizeof(void*));
+        if (list->data == NULL)
+        {
+            fprintf(stderr, "Error: list_resize: realloc failed\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+}
+
+void list_append(List* list, void* item)
+{
+    list->data[list->length++] = item;
+    list_resize(list);
```

Comment:
I think you want to resize before writing to the buffer; something like this:
```
list->length += 1
list_resize(list)
list->data[list->length - 1] = item
```

Otherwise, suppose the list is full (e.g. length=maxlength=16) and then you call append. You'll first write to memory that doesn't belong to you (data[16] is out of bounds if len(data)=16), and then resize to double the allocation to 32.